File: misc/002-Add Two Numbers/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        # Default exit case
        # If L1 is empty, then L2 is either empty, or is the return value.
        if(l1 == None):
            return l2
        # If L2 is empty then L1 was not empty, and is the return value.
        if(l2 == None):
            return l1

        # Create the head node for the return list
        l3, carry = self.addNode(l1,l2,0)
        n3 = l3
        l1 = l1.next if l1 is not None else None
        l2 = l2.next if l2 is not None else None

        while((l1 is not None) or (l2 is not None)):
            n3.next, carry = self.addNode(l1,l2,carry)

            # Step linked lists to next elems
            n3 = n3.next
            l1 = l1.next if l1 is not None else None
            l2 = l2.next if l2 is not None else None
        
        # Append carried value if applicable
        if carry is not 0:
            n3.next = ListNode(val=(carry), next=None)

        return l3

    def addNode(self, l1: Optional[ListNode], l2: Optional[ListNode], carry: Optional[int]=0) -> tuple[ListNode,int]:
        vL = 0 if l1 is None else l1.val
        vR = 0 if l2 is None else l2.val
        vTotal = vL + vR + carry
        vRemainder = vTotal % 10

        return(ListNode(val=vRemainder, next=None), vTotal//10)

File: misc/002-Add Two Numbers/main.c
#define NULL 0

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addNode(struct ListNode* l1, struct ListNode* l2, int* carry);

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2)
{
    // Base case
    // If L1 is empty, then L2 is either also empty, or contains solution.
    if(l1 == NULL)
        return l2;
    // If L2 is empty then L1 contains the solution.
    if(l2 == NULL)
        return l1;

    struct ListNode* l3, *w;
    int carry = 0;

    // Create return list head
    l3 = addNode(l1, l2, &carry);

    // Walk L1 and L2 to next values
    if(l1 != NULL)
        l1 = l1->next;
    
    if(l2 != NULL)
        l2 = l2->next;

    w = l3;
    while((l1 != NULL) || (l2 != NULL))
    {
        w->next = addNode(l1, l2, &carry);

        // Walk all three node lists
        w = w->next;

        if(l1 != NULL)
            l1 = l1->next;
    
        if(l2 != NULL)
            l2 = l2->next;
    }

    // Append straggling carried values
    if(carry)
    {
        w->next = malloc(sizeof(struct ListNode));
        w->next->val = carry;
        w->next->next = NULL;
    }

    return l3;

}

struct ListNode* addNode(struct ListNode* l1, struct ListNode* l2, int* carry)
{
    int v1, v2, vTotal, vRemainder;
    struct ListNode* rNode;

    // Initialize working data
    v1 = v2 = 0;
    if(l1 != NULL)
        v1 = l1->val;
    if(l2 != NULL)
        v2 = l2->val;

    // Calculate node values
    vTotal = v1 + v2 + *carry;
    vRemainder = vTotal % 10;
    *carry = vTotal / 10;

    // Instanticiate new node
    rNode = malloc(sizeof(struct ListNode));
    rNode->val = vRemainder;
    rNode->next = NULL;
    return rNode;
}